src/v/storage/segment_index.cc

```diff
@@ -177,27 +203,26 @@ ss::future<bool> segment_index::materialize_index() {
 
 ss::future<> segment_index::drop_all_data() {
     reset();
-    return _out.truncate(0);
+    co_return co_await (co_await open()).truncate(0);
 }
 
 ss::future<> segment_index::flush() {
     if (!_needs_persistence) {
         co_return;
     }
     _needs_persistence = false;
-    co_await _out.truncate(0);
-    auto out = co_await ss::make_file_output_stream(ss::file(_out.dup()));
+    auto backing_file = co_await open();
+    co_await backing_file.truncate(0);
+    auto out = co_await ss::make_file_output_stream(
+      ss::file(backing_file.dup()));
     auto b = serde::to_iobuf(_state.copy());
     for (const auto& f : b) {
         co_await out.write(f.get(), f.size());
     }
     co_await out.flush();
     co_await out.close();
 }
-ss::future<> segment_index::close() {
-    co_await flush();
-    co_await _out.close();
-}
+ss::future<> segment_index::close() { co_await flush(); }
```

Comment:
i still don't have a good mental model for the type changes here because the removal of close() here has me thinking. i guess i'm wondering if a different than than close would make it more intuitive?

Reply:
close() can just be removed: I've done that and replicated calls with a direct call to flush()